optuna/samplers/_tpe/probability_distributions.py

```diff
@@ -0,0 +1,136 @@
+from typing import List
+from typing import NamedTuple
+from typing import TYPE_CHECKING
+from typing import Union
+
+import numpy as np
+
+from optuna._imports import _LazyImport
+
+
+if TYPE_CHECKING:
+    import scipy.special as special
+    import scipy.stats as stats
+else:
+    special = _LazyImport("scipy.special")
+    stats = _LazyImport("scipy.stats")
+
+
+class _BatchedCategoricalDistributions(NamedTuple):
+    weights: np.ndarray
+
+
+class _BatchedTruncNormDistributions(NamedTuple):
+    mu: np.ndarray
+    sigma: np.ndarray
+    low: float  # Currently, low and high do not change per trial.
+    high: float
+
+
+class _BatchedDiscreteTruncNormDistributions(NamedTuple):
+    mu: np.ndarray
+    sigma: np.ndarray
+    low: float  # Currently, low, high and step do not change per trial.
+    high: float
+    step: float
+
+
+_BatchedDistributionUnion = Union[
+    _BatchedCategoricalDistributions,
+    _BatchedTruncNormDistributions,
+    _BatchedDiscreteTruncNormDistributions,
+]
+
+EPS = 1e-12
+
+
+def _normal_cdf(
+    x: Union[float, np.ndarray], mu: Union[float, np.ndarray], sigma: Union[float, np.ndarray]
+) -> np.ndarray:
+    return 0.5 * (1 + special.erf((x - mu) / np.maximum(np.sqrt(2) * sigma, EPS)))
+
+
+class _MixtureOfProductDistribution(NamedTuple):
```

Comment:
This class has 2 methods, so instead of making this a `NamedTuple`, it would be better to make it a regular class.

Reply:
@HideakiImamura Could you explain the reason why a `NamedTuple` with two methods would be bad?
@c-bata and I agreed that using `NamedTuple` would simplify the perceptual burden of reading the code. 
(There is another side effect of using `NamedTuple` that it's easier to compare two `_MixtureOfProductDistribution`.)

Reply:
I see. It is important that the `_MixtureOfProductDistribution` is immutable. Thanks. 

At first, a `NamedTuple` with a method seemed unnatural because it is only a named `Tuple`. I thought `NamedTuple` should behave as a container like `Dict`, but it is immutable. However, I think it is good because there is no inconvenience in using this form due to its great practical advantages.